Add Lokomotive concepts documentation (#111)
kinvolk/lokomotive: docs/concepts/components.md

# Lokomotive component

> `lokoctl` is a command line interface for Lokomotive and its *components*. A component is a

Reviewer: It's weird that this section has title "Lokomotive component" and it starts by talking about `lokoctl`...

Lokomotive cluster is just a barebone kubernetes cluster. In some cases (e.g. bare metal
environments), a Lokomotive cluster may even be unusable without certain components.

Components take care of tasks such as load balancing, monitoring, authentication, storage and
others.

## Categories of Lokomotive components

### User Authentication

Lokomotive provides the [Dex](../component-reference/components/dex.md) and
[Gangway](../component-reference/components/gangway.md) components for user authentication via
OpenID Connect (OIDC). With these components, you can securely manage access to the Kubernetes
cluster and resources.

Lokomotive also provides the [cert-manager](../component-reference/components/cert-manager.md)
component for automating the management and issuance of TLS certificates from various issuing
sources.

### Monitoring/Metrics

Lokomotive provides a
[prometheus-operator](../component-reference/components/prometheus-operator.md) component that
creates, configures and manages [Prometheus](https://prometheus.io/) atop Kubernetes.

Lokomotive also provides a [metrics-server](../component-reference/components/metrics-server.md)

> component responsible for collecting resource metrics from kubelets and exposing them in the

Reviewer: ```suggestion
component responsible for collecting resource metrics from nodes and pods and exposing them in the
```

Kubernetes API server through the [metrics API](https://github.com/kubernetes/metrics).

### Storage

Lokomotive provides the [openebs-operator](../component-reference/components/openebs-operator.md) and
[openebs-storage-class](../component-reference/components/openebs-storage-class.md) components to
use OpenEBS as the block storage solution for the cluster.

Lokomotive also provides the [rook](../component-reference/components/rook.md) and
[rook-ceph](../component-reference/components/rook-ceph.md) components for using Rook as the storage
solution for Lokomotive cluster.

### Load Balancing/Ingress (for Packet platform only)

Lokomotive provides the [MetalLB](../component-reference/components/metallb.md) component for load

> balancing in bare metal environments and [Contour](../component-reference/components/contour.md)

Reviewer: Saying bare metal environments may sound confusing here, as we also have bare-metal platform. Maybe we can re-phrase it to "in environments without programmatic load balancers?" or something?

> component for ingress.

Reviewer: ```suggestion
component for ingress control.
```

The suggestion is not optimal, but I would mention here, that Contour works as Ingress controller.

Lokomotive also provides the [external-dns](../component-reference/components/external-dns.md)
component for automatic management of DNS entries for Ingress resources.

### Node management

Lokomotive provides the
[calico-hostendpoint-controller](../component-reference/components/calico-hostendpoint-controller.md)
component for adding and removing [Calico HostEndpoint

> objects](https://docs.projectcalico.org/v3.8/reference/resources/hostendpoint) and the

Reviewer: ```suggestion
objects](https://docs.projectcalico.org/v3.8/reference/resources/hostendpoint) automatically and the
```

To install a Lokomotive component, run the following command:

```console
lokoctl component install <component_name>
```

> This will take the `kubeconfig` from your cluster asset directory if you run it in the folder of

Reviewer: I would rephrase it to something like:

> If this command is executed in the folder with cluster configuration, `lokoctl` will try to install the component on configured cluster. If no configuration is found, the configuration from `KUBECONFIG` environment variable will be used. If this one is empty, `~/.kube/config` file will be used.

>To use specific `kubeconfig` file, `--kubeconfig` flag can be used.
